data/transactions/verify/txn.go

```diff
@@ -314,3 +335,124 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
 	return nil
 
 }
+
+// PaysetGroups verifies that the payset have a good signature and that the underlying
+// transactions are properly constructed.
+// Note that this does not check whether a payset is valid against the ledger:
+// a PaysetGroups may be well-formed, but a payset might contain an overspend.
+//
+// This version of verify is performing the verification over the provided execution pool.
+func PaysetGroups(ctx context.Context, payset [][]transactions.SignedTxn, blk bookkeeping.Block, verificationPool execpool.BacklogPool) (err error) {
+	proto, ok := config.Consensus[blk.BlockHeader.CurrentProtocol]
+	if !ok {
+		return protocol.Error(blk.BlockHeader.CurrentProtocol)
+	}
+	if len(payset) == 0 {
+		return nil
+	}
+
+	// prepare up to 16 concurrent worksets.
+	worksets := make(chan struct{}, concurrentWorksets)
+	worksDoneCh := make(chan interface{}, concurrentWorksets)
+	processing := 0
+	tasksCtx, cancelTasksCtx := context.WithCancel(ctx)
+	defer cancelTasksCtx()
+	builder := worksetBuilder{payset: payset}
+	var nextWorkset [][]transactions.SignedTxn
+	for processing >= 0 {
+		// see if we need to get another workset
+		if len(nextWorkset) == 0 && !builder.completed() {
+			nextWorkset = builder.next()
+		}
+
+		select {
+		case <-tasksCtx.Done():
+			return tasksCtx.Err()
+		case worksets <- struct{}{}:
+			if len(nextWorkset) > 0 {
+				err := verificationPool.EnqueueBacklog(ctx, func(arg interface{}) interface{} {
+					// check if we've canceled the request while this was in the queue.
+					if tasksCtx.Err() != nil {
+						return tasksCtx.Err()
+					}
+					txnGroups := arg.([][]transactions.SignedTxn)
+					if !proto.SupportRekeying {
```

Comment:
Current
```
BenchmarkPaysetGroups-8   	   78248	     16482 ns/op
BenchmarkPaysetGroups-8   	   75268	     16444 ns/op

BenchmarkTxnPool-8   	   18794	     63580 ns/op
BenchmarkTxnPool-8   	   18135	     64093 ns/op
```

With vanila `Txn(...)` call and the only one two loops instead of 4:
```
					txnGroups := arg.([][]transactions.SignedTxn)
					for _, signTxnsGrp := range txnGroups {
						ctxs := PrepareContexts(signTxnsGrp, blk.BlockHeader)
						for k, signTxn := range signTxnsGrp {
							if err := Txn(&signTxn, ctxs[k]); err != nil {
								return err
							}
						}
					}
```
```
BenchmarkPaysetGroups-8   	   80364	     15178 ns/op
BenchmarkPaysetGroups-8   	   82190	     14846 ns/op

BenchmarkTxnPool-8   	   18781	     66758 ns/op
BenchmarkTxnPool-8   	   18808	     63886 ns/op
```

Reply:
I guess just calling `Txn` is desirable since it is not slower and gives twice less code and no code duplication.